opentelemetry-config/src/lib.rs

```diff
@@ -0,0 +1,323 @@
+        config: &dyn std::any::Any,
+    ) -> MeterProviderBuilder;
+
+    fn as_any(&self) -> &dyn std::any::Any;
+}
+
+/// Holds the configured telemetry providers
+pub struct TelemetryProviders {
+    meter_provider: Option<SdkMeterProvider>,
+    traces_provider: Option<SdkTracerProvider>,
+    logs_provider: Option<SdkLoggerProvider>,
+}
+
+impl TelemetryProviders {
+    pub fn new() -> Self {
+        TelemetryProviders {
+            meter_provider: None,
+            traces_provider: None,
+            logs_provider: None,
+        }
+    }
+
+    pub fn with_meter_provider(mut self, meter_provider: SdkMeterProvider) -> Self {
+        self.meter_provider = Some(meter_provider);
+        self
+    }
+
+    pub fn with_traces_provider(mut self, traces_provider: SdkTracerProvider) -> Self {
+        self.traces_provider = Some(traces_provider);
+        self
+    }
+    pub fn with_logs_provider(mut self, logs_provider: SdkLoggerProvider) -> Self {
+        self.logs_provider = Some(logs_provider);
+        self
+    }
+
+    pub fn meter_provider(&self) -> Option<&SdkMeterProvider> {
+        self.meter_provider.as_ref()
+    }
+
+    pub fn traces_provider(&self) -> Option<&SdkTracerProvider> {
+        self.traces_provider.as_ref()
+    }
+
+    pub fn logs_provider(&self) -> Option<&SdkLoggerProvider> {
+        self.logs_provider.as_ref()
+    }
+
+    pub fn shutdown(self) -> OTelSdkResult {
+        if let Some(meter_provider) = self.meter_provider {
+            meter_provider.shutdown()?;
+        }
+        if let Some(traces_provider) = self.traces_provider {
+            traces_provider.shutdown()?;
+        }
+        if let Some(logs_provider) = self.logs_provider {
+            logs_provider.shutdown()?;
+        }
+        Ok(())
+    }
+}
+
+impl Default for TelemetryProviders {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[derive(Debug)]
+pub enum ConfiguratorError {
+    InvalidConfiguration(String),
+    UnsupportedExporter(String),
+    NotRegisteredConfigurator(String),
+}
+
+impl error::Error for ConfiguratorError {}
+
+impl Display for ConfiguratorError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            ConfiguratorError::InvalidConfiguration(details) => {
+                write!(f, "Invalid configuration: {}", details)
+            }
+            ConfiguratorError::UnsupportedExporter(details) => {
+                write!(f, "Unsupported exporter: {}", details)
+            }
+            ConfiguratorError::NotRegisteredConfigurator(details) => {
+                write!(f, "Not registered configurator: {}", details)
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::sync::atomic::AtomicI16;
+
+    use super::*;
+
+    #[test]
+    fn test_register_periodic_exporter_configurator() {
+        // Arrange
+        struct MockPeriodicExporterConfigurator {
+            call_count: AtomicI16,
```

Comment:
These are single threaded tests. We don't need to use atomics here. You might want to use `Cell` instead.